pkg/log/zap/logger.go

```diff
@@ -75,6 +75,13 @@ func getConfig(destWriter io.Writer) config {
 	}
 	if levelVal.set {
 		c.level = zap.NewAtomicLevelAt(levelVal.level)
+
+		// Disable sampling when we are in debug mode. Otherwise, this will
+		// cause index out of bounds errors in the sampling code.
+		if levelVal.level < -1 {
+			sampleVal.set = false
+			c.sample = false
```

Comment:
Do we think it's okay to silently disable sampling when `--zap-sample=true`?

Instead, we could `log.Fatal()` or change the API to allow us to return an error. Thoughts?

Reply:
Yes, I do. 

We should document that this is what we do, but TBH how many users are purposefully turning on sampling (which is for log performance in production) AND setting the log level to trace level logs. I would assume this is a mistake and we should probably just make the right decision.

Reply:
Also, this is more of a nit, but I'm a little nervous about this check happening in the `if levelVal.set { ... }` block. I know we have unit tests now, but if we ever change our default log level to debug, this check would get skipped.

It would probably be better to use `levelVal.level` to set the default in line 69 and then do another check in the `if sampleVal.set { ... }` block to make sure we turn sampling back on when debug logging is enabled.

Reply:
@jmrodri I think we should use c.level to determine this on line 89.

Reply:
I moved the if block to out of the `levelVal.set` to just before return c.

```golang
func getConfig() config {                                                            
    var c config                                                                     
                                                                                     
    // Set the defaults depending on the log mode (development vs. production)       
    if development {                                                                 
        c.encoder = consoleEncoder()                                                 
        c.level = zap.NewAtomicLevelAt(zap.DebugLevel)                               
        c.opts = append(c.opts, zap.Development(), zap.AddStacktrace(zap.ErrorLevel))
        c.sample = false                                                             
    } else {                                                                         
        c.encoder = jsonEncoder()                                                    
        c.level = zap.NewAtomicLevelAt(zap.InfoLevel)                                
        c.opts = append(c.opts, zap.AddStacktrace(zap.WarnLevel))                    
        c.sample = true                                                              
    }                                                                                
                                                                                     
    // Override the defaults if the flags were set explicitly on the command line    
    if encoderVal.set {                                                              
        c.encoder = encoderVal.encoder                                               
    }                                                                                
    if levelVal.set {                                                                
        c.level = zap.NewAtomicLevelAt(levelVal.level)                               
                                                                                     
    }                                                                                
    if sampleVal.set {                                                               
        c.sample = sampleVal.sample                                                  
    }                                                                                
                                                                                     
    // Disable sampling when we are in debug mode. Otherwise, this will              
    // cause index out of bounds errors in the sampling code.                        
    if c.level.Level() < -1 {                                                        
        sampleVal.set = false                                                        
        c.sample = false                                                             
    }                                                                                
    return c                                                                         
}                                                                                    
```

Reply:
I moved the if block to just before we return the config.

```golang
func getConfig() config {                                                             
    var c config                                                                      
                                                                                      
    // Set the defaults depending on the log mode (development vs. production)        
    if development {                                                                  
        c.encoder = consoleEncoder()                                                  
        c.level = zap.NewAtomicLevelAt(zap.DebugLevel)                                
        c.opts = append(c.opts, zap.Development(), zap.AddStacktrace(zap.ErrorLevel)) 
        c.sample = false                                                              
    } else {                                                                          
        c.encoder = jsonEncoder()                                                     
        c.level = zap.NewAtomicLevelAt(zap.InfoLevel)                                 
        c.opts = append(c.opts, zap.AddStacktrace(zap.WarnLevel))                     
        c.sample = true                                                               
    }                                                                                 
                                                                                      
    // Override the defaults if the flags were set explicitly on the command line     
    if encoderVal.set {                                                               
        c.encoder = encoderVal.encoder                                                
    }                                                                                 
    if levelVal.set {                                                                 
        c.level = zap.NewAtomicLevelAt(levelVal.level)                                
                                                                                      
    }                                                                                 
    if sampleVal.set {                                                                
        c.sample = sampleVal.sample                                                   
    }                                                                                 
                                                                                      
    // Disable sampling when we are in debug mode. Otherwise, this will               
    // cause index out of bounds errors in the sampling code.                         
    if c.level.Level() < -1 {                                                         
        c.sample = false                                                              
    }                                                                                 
    return c                                                                          
}                                                                                     
```